src/memcpy_vfp.c

```diff
@@ -0,0 +1,151 @@
+        "   pld   [%1, #160]        \n\t"
+        "   pld   [%1, #192]        \n\t"
+        "   pld   [%1, #224]        \n\t"
+        "   pld   [%1, #256]        \n\t"
+        "   vldm  %1!, {d0-d15}     \n\t"
+        "   subs  %2, %2, #128      \n\t"
+        "   vstm  %0!, {d0-d15}     \n\t"
+        "   bgt   0b                \n\t"
+        : "+r" (dst), "+r" (src), "+r" (size)
+        :
+        : "cc", "d8", "d9", "d10", "d11", "d12", "d13", "d14", "d15");
+}
+
+void tegra_copy_block_vfp(char *dst, char *src, int size,
+                          bool bounce, bool move)
+{
+    int i, dir, block_size = BLOCK_SIZE;
+    char *tmpbuf = tmp_buf0;
+    char *psrc = src;
+    char *pdst = dst;
+
+    assert(((uintptr_t)src & 127) == 0);
+    assert(((uintptr_t)dst & 127) == 0);
+    assert((size & 127) == 0);
+
+    do {
+        if (size < block_size || !bounce) {
+            block_size = size;
+            move = false;
+        }
+
+        dir = (pdst > psrc && move) ? -1 : 1;
+
+        if (dir < 0) {
+            psrc += size - block_size;
+            pdst += size - block_size;
+        }
+
+        for (i = 0; i < size / block_size; i++) {
+            if (!bounce)
+                tmpbuf = pdst;
+
+            tegra_aligned_block_copy_vfp(tmpbuf, psrc, block_size);
+
+            if (bounce)
+                tegra_aligned_block_copy_vfp(pdst, tmpbuf, block_size);
+
+            psrc += block_size * dir;
+            pdst += block_size * dir;
+        }
+
+        size -= block_size * i;
+        pdst = max(pdst, dst);
+        psrc = max(psrc, src);
+
+    } while (size);
+}
+
+void tegra_do_memcpy_vfp(void *dst, const void *src, int size, bool bounce)
+{
+    bool src_aligned = !((uintptr_t)src & 127);
+    bool dst_aligned = !((uintptr_t)dst & 127);
+    int left = size;
+    int block_size;
+    char *csrc = (char *)src;
+    char *cdst = dst;
+    char *pdst;
+
+    if (src_aligned && dst_aligned && left > 127) {
+        tegra_copy_block_vfp(cdst, csrc, size & ~127, bounce, false);
+        csrc += size & ~127;
+        cdst += size & ~127;
+        left = size & 127;
+    }
+
+    if (left > 127) {
+        uintptr_t addr, addr_aligned, offset;
+
+        /* having dst bouncing instead of src is always more efficient */
+        addr         = (uintptr_t)csrc;
+        addr_aligned = TEGRA_ALIGN(addr, 128);
+        offset       = addr_aligned - addr;
+
+        if (offset) {
+            memcpy(cdst, csrc, offset);
+
+            csrc += offset;
+            cdst += offset;
+            left -= offset;
+
+            dst_aligned = !((uintptr_t)cdst & 127);
+        }
+    }
+
+    while (left > 127) {
+        block_size = min(BLOCK_SIZE, left & ~127);
+
+        if (!dst_aligned)
+            pdst = tmp_buf0;
+        else
+            pdst = cdst;
+
+        tegra_copy_block_vfp(pdst, csrc, block_size, false, false);
+
+        if (!dst_aligned)
+            memcpy(cdst, tmp_buf0, block_size);
```

Comment:
I'm having a bit trouble believing that `tegra_copy_block_vfp` + `memcpy` will be faster than straight `memcpy` in the case where alignment differs. In fact, I don't really understand why we need the whole `tegra_copy_block_vfp` in the first place...

Wouldn't something like this be simpler and maybe faster?
```C
void tegra_do_memcpy_vfp(void *dst, const void *src, int size)
{
   uint8_t *cdst;
   const uint8_t *csrc;
   int align = (uintptr_t)src & 127;
   if ((uintptr_t)dst & 127 != align || size < 128) {
      memcpy(dst, src, size);
      return;
   }

   /* align both pointers */
   align = 128 - align;
   memcpy(dst, src, dst_align);
   cdst = (uint8_t *)dst + align;
   csrc = (const uint8_t *)src + align;
   size -= align;

   if (size > 128) {
      /* copy aligned data */
      int block_size = size & ~127;
      tegra_aligned_block_copy_vfp(pdst, psrc, block_size);
      csrc += block_size;
      cdst += block_size;
      size -= block_size;
   }

   /* copy trailing data */
   if (size > 0)
      memcpy(cdst, csrc, size);
}
```

Reply:
In our case the dst is either cached or write-combined, hence CPU can help with re-aligning and combining of the written data.

Generic reading of uncached memory is always slow, hence we'are aligning the src data so that it can be read by VFP. Gllib's memcpy() has lot's of code for optimizing unaligned copies and the rest, it also assumes that read data is prefetchable, that all doesn't work for uncached readings. Also, my understanding is that memory BUS is 64bit wide, but ARM32 can't read raw memory word at once without using VFP instruction. Moreover VFP reads 128bytes at a time, it certainly should benefit reading/writing in a single burst and large-sequential memory accesses are more optimal in HW.

Both pointers can't be aligned if one of them is unaligned.

Reply:
Maybe I wasn't so clear. The way I read the code, you end up pulling the *all* data through memcpy if not both pointers are aligned. So what's the gain?

Reply:
That may happen only if copied size is very small, so that after aligning there is less than 128B to copy.. are you taking about this case?

If dst is unaligned, then uncached data is copied to the cacheable bounce buffer and then memcpy'ed to dst. This works very efficiently, probably because of the L2 caching and prefetching that are helping a lot. Maybe this case could be optimized further a tad.

Reply:
No, it happens whenever src and dst have different alignment-distances AFAICT. If the distances isn't the same, dst_aligned will be false when it's re-evaluated.

But OK, so the point is that `tegra_copy_block_vfp` into a temp buffer + `memcpy` into the target is faster than just plain memcpy? That sounds a bit hard to believe, copying the same data twice means twice the memory bandwidth consumed...

Even an LDMIA/STMIA loop can manage this very well on uncached memory as long as both buffers are 4 byte aligned. I can almost guarantee that's going to be faster than two copies... And if we can't guarantee *4 bytes* alignment in practice, I think somethings very broken somewhere...

Reply:
`ARM copy (from framebuffer)` is LDMIA/STMIA loop, see perf difference in comparison with VFP.

I just tried to enforce the bounce-buffering for the dst.. and very interestingly this VFP + memcpy combination makes copying faster by 40 MB/s in comparison to direct VFP copying.

Copying same data twice doesn't mean twice the memory bandwidth if it is hitting cache and not the memory.

Reply:
Right, I suppose I was getting a bit stuck in the "uncached"-bit, where it definitely would have been twice. But since the bounce-buffer is cached, I suppose this could make some sense.

I'm trying to run the same test here, but it seems like tinymembench doesn't want to run the Framebuffer tests here at all... hmmpf...

Reply:
Ah, sudo :)

Reply:
Perhaps there is a room for more optimizations here.